csaf/src/metadata_retriever_new/mod.rs

```diff
@@ -0,0 +1,259 @@
+
+            let dns_result = self
+                .fetcher
+                .fetch::<Json<ProviderMetadata>>(dns_path.clone())
+                .await?;
+            Ok(DetectionType::DnsPath(dns_result.into_inner()))
+        } else {
+            Err(Error::Url(ParseError::EmptyHost))
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl Retriever for SecurityTextRetriever {
+    async fn retrieve(&self) -> Result<DetectionType, Error>
+    where
+        Self: Sized + Send,
+    {
+        log::trace!(
+            "Starting retrieve provider_metadata from security text of provided discovery URL "
+        );
+        if let Some(host_url) = self.provider_metadata_url.clone().host_str() {
+            let host_url_string = format!("https://{}", &host_url);
+            let host_url = Url::parse(host_url_string.as_str())?;
+            log::warn!(
+                "Querying provider metadata url from security text of {:?}",
+                host_url_string
+            );
+            let provider_metadata_path =
+                get_url_from_security_text(&self.fetcher, host_url.join("/security.txt")?).await?;
+
+            let security_url = Url::parse(provider_metadata_path.as_str())?;
+            let security_result = self
+                .fetcher
+                .fetch::<Json<ProviderMetadata>>(security_url.clone())
+                .await?;
+            Ok(DetectionType::SecurityTextPath(
+                security_result.into_inner(),
+            ))
+        } else {
+            Err(Error::Url(ParseError::EmptyHost))
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl Retriever for FullSecurityTextRetriever {
+    async fn retrieve(&self) -> Result<DetectionType, Error>
+    where
+        Self: Sized + Send,
+    {
+        log::trace!(
+            "Starting retrieve provider_metadata from security text of provided discovery URL "
+        );
+        if let Some(host_url) = self.provider_metadata_url.clone().host_str() {
+            let host_url_string = format!("https://{}", &host_url);
+            let host_url = Url::parse(host_url_string.as_str())?;
+            log::warn!(
+                "Querying provider metadata url from security text of {:?}",
+                host_url_string
+            );
+            let provider_metadata_path = get_url_from_security_text(
+                &self.fetcher,
+                host_url.join("/.well-known/security.txt")?,
+            )
+            .await?;
+            let security_url = Url::parse(provider_metadata_path.as_str())?;
+            let security_result = self
+                .fetcher
+                .fetch::<Json<ProviderMetadata>>(security_url.clone())
+                .await?;
+            Ok(DetectionType::FullSecurityTextPath(
+                security_result.into_inner(),
+            ))
+        } else {
+            Err(Error::Url(ParseError::EmptyHost))
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl Retriever for AllApproachRetriever {
```

Comment:
Assuming each of those functions would return `None` in the case the function could not detect the metadata (still completed successfully), then I think we should chain this using `or_else` (or one of those functions).